app/backend/src/tests/test_donations.py

```diff
@@ -0,0 +1,259 @@
+
+    new_config = config.copy()
+    new_config["ENABLE_DONATIONS"] = True
+    new_config["STRIPE_API_KEY"] = "dummy_api_key"
+    new_config["STRIPE_WEBHOOK_SECRET"] = "dummy_webhook_secret"
+    new_config["STRIPE_RECURRING_PRODUCT_ID"] = "price_1IRoHdE5kUmYuPWz9tX8UpRv"
+
+    monkeypatch.setattr(couchers.servicers.donations, "config", new_config)
+
+    ## User first makes a req to Donations.InitiateDonation
+    with donations_session(token) as donations:
+        with patch("couchers.servicers.donations.stripe") as mock:
+            mock.Customer.create.return_value = type("__MockCustomer", (), {"id": "cus_JoeqxRiXxWnThy"})
+            mock.checkout.Session.create.return_value = type(
+                "__MockCheckoutSession",
+                (),
+                {
+                    "id": "cs_test_a1ScieS4Qtnak4fI7qhUChYz0XBBzwRBKMjNESus7eiF0WvIcV07i3TFx5",
+                    "payment_intent": "pi_1JB4HaE5kUmYuPWz9O9m5vd9",
+                },
+            )
+
+            res = donations.InitiateDonation(
+                donations_pb2.InitiateDonationReq(
+                    amount=100,
+                    recurring=False,
+                )
+            )
+
+        mock.Customer.create.assert_called_once_with(
+            email=user_email, metadata={"user_id": user_id}, api_key="dummy_api_key"
+        )
+
+        mock.checkout.Session.create.assert_called_once_with(
+            client_reference_id=user_id,
+            submit_type="donate",
+            customer="cus_JoeqxRiXxWnThy",
+            success_url="http://localhost:3000/donate?success=true",
+            cancel_url="http://localhost:3000/donate?cancelled=true",
+            payment_method_types=["card"],
+            mode="payment",
+            line_items=[
+                {
+                    "price_data": {
+                        "currency": "usd",
+                        "unit_amount": 10000,
+                        "product_data": {
+                            "name": f"Couchers financial supporter (one-time)",
+                            "images": ["https://couchers.org/img/share.jpg"],
+                        },
+                    },
+                    "quantity": 1,
+                }
+            ],
+            api_key="dummy_api_key",
+        )
+
+    ## Stripe then makes some webhooks requests
+
+    # evt_1JB4HbE5kUmYuPWzIhui6KMm: payment_intent.created
+    fire_stripe_event("evt_1JB4HbE5kUmYuPWzIhui6KMm")
+
+    # evt_1JB4HyE5kUmYuPWzOgFnJ83g: payment_intent.succeeded
+    fire_stripe_event("evt_1JB4HyE5kUmYuPWzOgFnJ83g")
+
+    # evt_1JB4HyE5kUmYuPWzATTZL7Mt: charge.succeeded
+    fire_stripe_event("evt_1JB4HyE5kUmYuPWzATTZL7Mt")
+
+    # evt_1JB4HyE5kUmYuPWzINvGWlRu: checkout.session.completed
+    fire_stripe_event("evt_1JB4HyE5kUmYuPWzINvGWlRu")
+
+    ## Now finally check everything was added to the DB
+    with session_scope() as session:
+        donation = session.execute(select(OneTimeDonation)).scalar_one()
```

Comment:
Should we add a where? Better to be more robust/precise, especially protects against the test changing.

Reply:
This is already robust to there being exactly one donation (it'll raise an exception and fail if there's none or more than one), I think that's enough: we then assert that it's the correct one.